pkg/beholder/auth.go

```diff
@@ -2,26 +2,75 @@ package beholder
 
 import (
 	"crypto/ed25519"
+	"encoding/binary"
 	"fmt"
+	"time"
 )
 
 // authHeaderKey is the name of the header that the node authenticator will use to send the auth token
 var authHeaderKey = "X-Beholder-Node-Auth-Token"
 
 // authHeaderVersion is the version of the auth header format
-var authHeaderVersion = "1"
+var authHeaderVersion1 = "1"
+var authHeaderVersion2 = "2"
+
+// AuthHeaderConfig is a configuration struct for the BuildAuthHeadersV2 function
+type AuthHeaderConfig struct {
+	timestamp int64
+	version   string
+}
 
 // BuildAuthHeaders creates the auth header value to be included on requests.
-// The current format for the header is:
+// There are two formats for the header. Version `1` is:
 //
 // <version>:<public_key_hex>:<signature_hex>
 //
 // where the byte value of <public_key_hex> is what's being signed
+// and <signature_hex> is the signature of the public key.
+// The version `2` is:
+//
+// <version>:<public_key_hex>:<timestamp>:<signature_hex>
+//
+// where the byte value of <public_key_hex> and <timestamp> are what's being signed
 func BuildAuthHeaders(privKey ed25519.PrivateKey) map[string]string {
 	pubKey := privKey.Public().(ed25519.PublicKey)
 	messageBytes := pubKey
 	signature := ed25519.Sign(privKey, messageBytes)
-	headerValue := fmt.Sprintf("%s:%x:%x", authHeaderVersion, messageBytes, signature)
 
-	return map[string]string{authHeaderKey: headerValue}
+	return map[string]string{authHeaderKey: fmt.Sprintf("%s:%x:%x", authHeaderVersion1, messageBytes, signature)}
+}
+
+func BuildAuthHeadersV2(privKey ed25519.PrivateKey, config *AuthHeaderConfig) map[string]string {
+	if config == nil {
+		config = defaultAuthHeaderConfig()
+	}
+	if config.version == "" {
+		config.version = authHeaderVersion2
+	}
+	// If timestamp is not set, use the current time
+	if config.timestamp == 0 {
+		config.timestamp = time.Now().UnixMilli()
+	}
+	// If timestamp is negative, set it to 0. negative values cause overflow on convertsion to uint64
+	// 0 timestamps will be rejected by the server as being too old
+	if config.timestamp < 0 {
+		config.timestamp = 0
```

Comment:
this sorta conflicts in spirit to the previous block where we set the timestamp to `now` if it's `0`

should this be moved up?

Reply:
So you're saying if a negative value is passed in, we should auto correct to current time?

Reply:
I think let's keep as is then